File: app/src/main/java/com/harishsk/saveblood/SplashScreen.java
package com.harishsk.saveblood;

import android.annotation.TargetApi;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.Set;

public class SplashScreen extends AppCompatActivity {

    private static final String TAG = SplashScreen.class.getSimpleName();

    private static final int PERMISSIONS_REQUEST = 1234;

    private Boolean bdelay = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash_screen);

        if (Build.VERSION.SDK_INT >= 23) checkPermissions();

        int delay = 1000;
        if (bdelay) delay = 3000;
        if (!bdelay)
            new Handler().postDelayed(this::main, delay);
        else
            new Handler().postDelayed(this::main, 6000);
    }

    @TargetApi(Build.VERSION_CODES.M)
    private void checkPermissions() {
        String[] ungrantedPermissions = requiredPermissionsStillNeeded();
        if (ungrantedPermissions.length != 0) {
            requestPermissions(ungrantedPermissions, PERMISSIONS_REQUEST);
            bdelay = true;
        }
    }

    public String[] getRequiredPermissions() {
        String[] permissions = null;
        try {
            permissions = getPackageManager().getPackageInfo(getPackageName(),
                    PackageManager.GET_PERMISSIONS).requestedPermissions;
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        if (permissions == null) return new String[0];
        else return permissions.clone();
    }

    @TargetApi(23)
    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        if (requestCode == PERMISSIONS_REQUEST) checkPermissions();
    }

    @TargetApi(23)
    private String[] requiredPermissionsStillNeeded() {

        Set<String> permissions = new HashSet<>();
        Collections.addAll(permissions, getRequiredPermissions());
        for (Iterator<String> i = permissions.iterator(); i.hasNext(); ) {
            String permission = i.next();
            if (checkSelfPermission(permission) == PackageManager.PERMISSION_GRANTED) {
                Log.d(TAG, getString(R.string.Permission) + permission + getString(R.string.already_granted));
                i.remove();
            } else
                Log.d(TAG, getString(R.string.Permission) + permission + getString(R.string.not_granted));
        }
        return permissions.toArray(new String[permissions.size()]);
    }

    public void main() {
        startActivity(new Intent(SplashScreen.this, PhoneAuthActivity.class));
        finish();
    }
}

File: app/src/main/java/com/harishsk/saveblood/ui/SplashScreen.kt
package com.harishsk.saveblood.ui

import android.annotation.TargetApi
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import com.harishsk.saveblood.R
import java.util.*

class SplashScreen : AppCompatActivity() {

    private var bdelay = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash_screen)
        if (Build.VERSION.SDK_INT >= 23) checkPermissions()
        var delay = 1000
        if (bdelay) delay = 3000
        when {
            !bdelay -> Handler(mainLooper).postDelayed(
                { main() },
                delay.toLong()
            )
            else -> Handler(mainLooper).postDelayed(
                { main() }, 6000
            )
        }
    }

    @TargetApi(Build.VERSION_CODES.M)
    private fun checkPermissions() {
        val unGrantedPermissions = requiredPermissionsStillNeeded()
        if (unGrantedPermissions.isNotEmpty()) {
            requestPermissions(
                unGrantedPermissions,
                PERMISSIONS_REQUEST
            )
            bdelay = true
        }
    }

    private val requiredPermissions: Array<String?>
        get() {
            var permissions: Array<String?>? = null
            try {
                permissions = packageManager.getPackageInfo(
                    packageName,
                    PackageManager.GET_PERMISSIONS
                ).requestedPermissions
            } catch (e: PackageManager.NameNotFoundException) {
                e.printStackTrace()
            }
            return permissions?.clone() ?: arrayOfNulls(0)
        }

    @TargetApi(23)
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        if (requestCode == PERMISSIONS_REQUEST) checkPermissions()
    }

    @TargetApi(23)
    private fun requiredPermissionsStillNeeded(): Array<String> {
        val permissions: MutableSet<String> =
            HashSet()
        Collections.addAll<String>(permissions, *requiredPermissions)
        val i = permissions.iterator()
        while (i.hasNext()) {
            val permission = i.next()
            when (PackageManager.PERMISSION_GRANTED) {
                checkSelfPermission(permission) -> {
                    Log.d(
                        TAG,
                        getString(R.string.Permission) + permission + getString(
                            R.string.already_granted
                        )
                    )
                    i.remove()
                }
                else -> Log.d(
                    TAG,
                    getString(R.string.Permission) + permission + getString(
                        R.string.not_granted
                    )
                )
            }
        }
        return permissions.toTypedArray()
    }

    private fun main() {
        startActivity(Intent(this@SplashScreen, PhoneAuthActivity::class.java))
        finish()
    }

    companion object {
        private val TAG =
            SplashScreen::class.java.simpleName
        private const val PERMISSIONS_REQUEST = 1234
    }
}